### ISSUE TITLE
GNN in ONNX: InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running ScatterElements node. Name:'ScatterElements_26' Status Message: Indices dim=12 at pos=0 is greater than input dim=6

### ISSUE THREAD

**Issue by Shra1-25** [onnx/onnx#3711]
# Ask a Question

### Question
Hello Everyone, So past few days I have been trying to run GNN using ONNX runtime (Python and eventually on C++). However I get the following error after I successfully export the onnx model from pytorch which uses pytorch geometric layers (SAGE Convolution to be precise) and functions for the GNNs. 

`InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running ScatterElements node. Name:'ScatterElements_26' Status Message: Indices dim=12 at pos=0 is greater than input dim=6` 

### Further information
Here is the code I used for pytorch model creation:
```python
#%debug
import torch
import torchvision
import torch_geometric
import torch.nn.functional as F
from torch.nn import Linear as Lin, Sequential as Seq, ReLU, AvgPool1d, Dropout, Softmax, Sigmoid, BatchNorm1d
from torch_geometric.nn import EdgeConv, DynamicEdgeConv, BatchNorm, avg_pool_x, knn_graph, EdgePooling, DenseSAGEConv, dense_diff_pool, dense_mincut_pool, SAGEConv, max_pool_x, SAGPooling, JumpingKnowledge, global_mean_pool as gap, global_max_pool as gmp
#from torch_cluster import knn_graph

#dummy_input = (torch.randn(10,5, device='cuda'),torch.randn(2,10, device='cuda').type(torch.int64))
coords0 = torch.randn(1,6)
coords1 = torch.randn(1,6)


coords = torch.transpose(torch.cat((coords0,coords1),dim=0),0,1)
adj = knn_graph(coords, k=2, batch=None, loop=True)

edge_from = adj[0:1,:]
edge_to = adj[1:,:]

input_values = (coords0,coords1, edge_from, edge_to)
#model = torchvision.models.alexnet(pretrained=True).cuda()
class trial1(torch.nn.Module):
    def __init__(self):
        super(trial1, self).__init__()
        
        self.SAGEConvBlock1 = SAGEConv(2, 512, normalize=True)
        self.bano1 = BatchNorm(512)
        self.relu = ReLU()
        self.dense1 = Seq(Lin(512,1))
        self.sigmoid = Sigmoid()
    def forward(self, coords0, coords1, edge_from, edge_to):
      adj = torch.cat((edge_from,edge_to),dim=0)
      gra = torch.transpose(torch.cat((coords0,coords1),dim=0),0,1)
      
      x1 = self.SAGEConvBlock1(gra, edge_index=adj)
      x = torch.unsqueeze(torch.sum(x1), dim=0)
      return x

model2 = trial1()

input_names = ["coords0","coords1", "edge_from", "edge_to"]
output_names = [ "outputs" ]

torch.onnx.export(model2, input_values, "ONNX_models/trial2.onnx", verbose=True, input_names=input_names, output_names=output_names, opset_version=11, dynamic_axes={'coords0':{0:'batch_size',1:'features'}, 'coords1':{0:'batch_size',1:'features'}, 'edge_from':{0:'batch_size',1:'features'}, 'edge_to':{0:'batch_size',1:'features'}, 'outputs':{0:'batch_size'}})
```

The onnx runtime code to perform the inference is:
```python
import onnxruntime
print(onnxruntime.__version__)
ort_session = onnxruntime.InferenceSession("ONNX_models/trial2.onnx")

def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(coords0), ort_session.get_inputs()[1].name: to_numpy(coords1), ort_session.get_inputs()[2].name: to_numpy(edge_from), ort_session.get_inputs()[3].name: to_numpy(edge_to)}
ort_outs = ort_session.run(["outputs"], ort_inputs)
``` 

I get the error in the last line of the above code: 
`ort_outs = ort_session.run(["outputs"], ort_inputs)`

Also, there is no error and the inference runs perfectly fine when I remove the SAGEConv layer from the model. The technical issue is the dimension mismatch between gra (shape: [6,2] )  and adj (shape: [2,36] )  where apparently the gra.shape[0] should be equal to adj.shape[1]. And it indeed works if we use only the first 6 elements of the adj (shape: [2,6] ) which essentially is not what we would intend to do. For my application I need to have multiple inputs of shape ('batch_size',nodes) and edges of the shape ('batch_size', k*nodes) where k indicates the number of k nearest neighbours.

### Notes
I would really appreciate if someone has any leads on this or if anyone has successfully deployed a GNN on ONNX using different input dimensions (rank <= 2).

**Comment by jcwchen** [comment]
Hi @Shra1-25,
Since the converted model can not work with different dimensional inputs, it seems to me that something wrong during the conversion or Runtime. Do you think the model is converted as expected? (It should not assume all input dimensions are the same; especially the original PyTorch model are not) I would suggest you first raise an issue in the converter (inside PyTorch repo) to make sure the graph is correctly converted and ask the converter experts whether SAGE Convolution is fully supported by the converter now. If so, then there might be something wrong in the Runtime (ONNXRuntime). Close the issue here. Please reopen it if you still have ONNX questions. Thank you!

**Comment by Shra1-25** [comment]
> Hi @Shra1-25,
> Since the converted model can not work with different dimensional inputs, it seems to me that something wrong during the conversion or Runtime. Do you think the model is converted as expected? (It should not assume all input dimensions are the same; especially the original PyTorch model are not) 

Yes the exported onnx model has correct input dimensions. On printing the following line, I get the correct shapes of the inputs. 
`ort_session.get_inputs()[0].shape, ort_session.get_inputs()[1].shape, ort_session.get_inputs()[2].shape, ort_session.get_inputs()[3].shape, ort_session.get_outputs()[0].shape`

To clarify, the issue is not with using different dimensional inputs, since when I remove the SAGEConv layer, I am able to run/execute the model with different input dimensions. It is only when I add the SAGEConv layer I get the error.

**Comment by jcwchen** [comment]
Thank you for the quick response and clarification. IIUC, the model can be run only by same input dimensions (1, 6) for `coords0, coords1, edge_from, edge_to`, but actually the original graph is using input dimensions: (1, 6), (1, 6), (1, 12), (1, 12). That's why I guess there might be something wrong during the conversion because the converted model only works with the same input dimensions, which is totally not the case of original PyTorch model...

To confirm that, you can use netron to visualize the ONNX graph and try to go through the nodes until ScatterElements node and see whether the model indeed requires the input dimensions are identical. If so, the conversion seems wrong (original PyTorch model does not require same input dimensions). If not, then the issue might be in ONNXRuntime instead (produced wrong indices for ScatterElements). Thanks
